cmake/CMakeLists.txt

```diff
@@ -9,7 +9,8 @@ IF(VERSION_STAGE)
 	SET(CPACK_PACKAGE_VERSION_PATCH         "${CPACK_PACKAGE_VERSION_PATCH}-${VERSION_STAGE}")
 ENDIF()
 IF(VERSION_BUILD)
-	SET(CPACK_PACKAGE_VERSION_PATCH         "${CPACK_PACKAGE_VERSION_PATCH}.${VERSION_BUILD}")
+	math(EXPR CPACK_PACKAGE_VERSION_PATH_BUMP "${CPACK_PACKAGE_VERSION_PATH}+1")
```

Comment:
Minor: `math` versus `MATH` above is inconsistent formatting.

Why bump this again for cpack?  Note, IIRC cpack is only used for Windows when `make package` is invoked; whereas Linux and macOS use shell scripts to name the binaries so this may result in inconsistently numbered binaries.   I'm still confused by the `+1` and then a subsequent `+1`.  I also thought the `+N` was for indicating commits and was already being accounted for, but I'm probably remembering the conversation on Discord incorrectly.

Reply:
The double bump is entirely because I have a poor understanding of how all the different CMake files interact. I'll make sure to fix the inconsistent case in my next commit, good point.

Can you elaborate on this part? I'm not sure I understand:

> I also thought the +N was for indicating commits and was already being accounted for, but I'm probably remembering the conversation on Discord incorrectly.

Reply:
> Can you elaborate on this part? I'm not sure I understand:
> > I also thought the +N was for indicating commits and was already being accounted for, but I'm probably remembering the conversation on Discord incorrectly.

Put simply, what was the point of bumping this? For example, if `stable-1.2` in july was technically `stable-1.2.1+57` by commit count calculation, why `+1`?  What is this `+1` for?  Shouldn't we let the VCS determine that (not some arbitrary `+1` logic?)

Reply:
Because prerelease versions are less than release versions*, and adding a commit count can only be done with prerelease identifiers**. If the latest tag is 1.2.2, any commits after that have to be bigger than 1.2.2, hence they must at least be a 1.2.3 prerelease.

*Clarification: according to semver, 1.2.2-500 < 1.2.2
**1.2.2.500 is an invalid semver version. It's unreasonable to increment the patch version for every commit, so we have to use prerelease labels

Reply:
> The double bump is entirely because I have a poor understanding of how all the different CMake files interact. I'll make sure to fix the inconsistent case in my next commit, good point.

Ah.. well, if it was needed for Windows, it should be warranted.  IIRC, Linux and Mac will just use what's set in the main CMakeLists.txt so they should be OK, I'm just surprised to see that CPack needed this.

Reply:
> Ah.. well, if it was needed for Windows, it should be warranted. IIRC, Linux and Mac will just use what's set in the main CMakeLists.txt so they should be OK, I'm just surprised to see that CPack needed this.

Pardon my ignorance, but just to clarify: The goal is for windows, mac, and linux to all have the same version number. Will the current logic accomplish this? If not, what should I change?

Reply:
@Spekular order of operations shows that `ADD_SUBDIRECTORY(cmake)` (~L616) comes after `VERSION_RELEASE_BUMP` (~L47) however, the value appears to be set on `VERSION` variable, not `VERSION_PATCH`, so I believe it's a consistency issue with the PR, not necessarily a logical issue with CMake.

https://github.com/LMMS/lmms/blob/6a73c281648c31d171bd9f7fc56f53bf85e84917/CMakeLists.txt#L47

https://github.com/LMMS/lmms/blob/6a73c281648c31d171bd9f7fc56f53bf85e84917/CMakeLists.txt#L616 

CPack uses its own variable for this and some unobvious NSIS integration uses it in the Windows installers (logic I fully plan to remove at some future time) but so as long as we're using it, we should keep it consistent.

https://github.com/LMMS/lmms/blob/6a73c281648c31d171bd9f7fc56f53bf85e84917/cmake/CMakeLists.txt#L7

My recommendation is to do one of the following:

1. Fix `VERSION_RELEASE` directly, get rid of `VERSION_RELEASE_BUMP` variable so it flows downstream from there.<br>
-- OR --
2. Tell CPack to use `VERSION_RELEASE_BUMP` and remove the redundant calculation.


> Will the current logic accomplish this

I haven't tested it, but at a glance, the redundant code will probably work at the expense of duplicated logic.

Reply:
> My recommendation is to do one of the following:
> 
> 1. Fix `VERSION_RELEASE` directly, get rid of `VERSION_RELEASE_BUMP` variable so it flows downstream from there.
>    -- OR --
> 2. Tell CPack to use `VERSION_RELEASE_BUMP` and remove the redundant calculation.

So `lmms/cmake/CMakeLists` can access variables that are set inside `lmms/CMakeLists`? (And it (`lmms/cmake/CMakeLists`) executes after `lmms/CMakeLists`?) If that's the case, I think I'll go with option one. Thanks for you patience in explaining this :)

Reply:
> So `lmms/cmake/CMakeLists` can access variables that are set inside `lmms/CMakeLists`? 

Correct, that's how it gets access to `VERSION_RELEASE`.  There's an order of operations that must occur, but that's been happening for a long time and is unlikely to change, quoting:

> `ADD_SUBDIRECTORY(cmake)` (~L616) comes after `VERSION_RELEASE_BUMP` (~L47) (or `VERSION_RELEASE`, etc)

If `ADD_SUBDIRECTORY(cmake)` where done prior to setting that variable, the variable would never propagate.  This is  why defining version information so early in the CMakeLists.txt is so important.